### PR TITLE
Generalize http fetch

### PR DESCRIPTION
**Pull request by skuzins**
- allow bytes as request body
- expose request and response headers in API
- update http example to show response headers and allow POST requests

Closes <https://github.com/emilk/egui/issues/487>.

### CODE REVIEW

File: epi/src/lib.rs [hunk 1]
```diff
@@ -377,34 +377,50 @@ pub const APP_KEY: &str = "app";
...
     pub struct Request {
         /// "GET", …
         pub method: String,
         /// https://…
         pub url: String,
-        /// x-www-form-urlencoded body
-        pub body: String,
+        /// The raw bytes.
+        pub body: Vec<u8>,
+        // ("Accept", "*/*"), …
```
Comment (emilk) [inline epi/src/lib.rs:390]:
```suggestion
        /// ("Accept", "*/*"), …
```

**Comment by emilk** [comment]
this looks good!

File: epi/src/lib.rs (hunk 1 above)
Comment (skuzins) [reply epi/src/lib.rs]: (made the change outside of this conversation, sorry)

**Comment by skuzins** [comment]
thanks, I'd say it's good to go

File: epi/src/lib.rs [hunk 2]
```diff
@@ -419,18 +435,34 @@ pub mod http {
...
+            } else {
+                None
+            }
+            */
+        }
+        pub fn content_type(&self) -> Option<String> {
+            self.headers.get("content-type").cloned()
+        }
     }
-
```
Comment (emilk) [inline epi/src/lib.rs:433]: please keep the blank lines around `impl` blocks etc

File: epi/src/lib.rs [hunk 3]
```diff
@@ -419,18 +435,34 @@ pub mod http {
...
         /// The raw bytes.
         pub bytes: Vec<u8>,
 
-        /// UTF-8 decoded version of bytes.
-        /// ONLY if `header_content_type` starts with "text" and bytes is UTF-8.
-        pub text: Option<String>,
+        pub headers: BTreeMap<String, String>,
+    }
+    impl Response {
```
Comment (emilk) [inline epi/src/lib.rs:443]:
```suggestion
    }
    
    impl Response {
```

File: epi/src/lib.rs [hunk 4]
```diff
@@ -419,18 +435,34 @@ pub mod http {
...
+            let content_type = self.content_type().unwrap_or_default();
+
+            if content_type.starts_with("text")
+                || content_type == "application/javascript"
+            {
+                String::from_utf8(self.bytes.clone()).ok()
+            } else {
+                None
+            }
+            */
```
Comment (emilk) [inline epi/src/lib.rs:460]:
```suggestion
```

File: epi/src/lib.rs [hunk 5]
```diff
@@ -419,18 +435,34 @@ pub mod http {
...
+        }
+        pub fn content_type(&self) -> Option<String> {
```
Comment (emilk) [inline epi/src/lib.rs:462]:
```suggestion
        }
        
        pub fn content_type(&self) -> Option<String> {
```

**Review by emilk — CHANGES_REQUESTED** [review]
Looks good, just some minor formatting nits!

And thank you!

**Review by emilk — APPROVED** [review]
Thanks!

**Comment by emilk** [comment]
Some cargo fmt errors. Tip: configure your editor to run `cargo fmt` on save.

**Comment by skuzins** [comment]
Thanks again @emilk for those suggestions. Anything else to do/fix on my end?

**Pull request merged**